### PR TITLE
use same HTTP client as dockerengine

### PR DESCRIPTION
**Pull request by vieux**
This should reduce the number of open connections

Close https://github.com/docker/swarm/issues/1549

### CODE REVIEW

**Comment by vieux** [comment]
ping @dongluochen

File: api/handlers.go [hunk 1]
```diff
@@ -130,7 +130,7 @@ func getImages(c *context, w http.ResponseWriter, r *http.Request) {
 	}
 
 	// Look for an engine that has all the images we need.
-	for engine, images := range engineImages {
+	for _, images := range engineImages {
```
Comment (MHBauer) [inline api/handlers.go]: If we're not going to use the first argument, can we avoid constructing it up above on L#127? Or maybe put `image.Engine` in there as the key.

File: api/utils.go [hunk 2]
```diff
@@ -102,14 +96,14 @@ func closeIdleConnections(client *http.Client) {
...
 }
 
-func proxyAsync(tlsConfig *tls.Config, addr string, w http.ResponseWriter, r *http.Request, callback func(*http.Response)) error {
-	// Use a new client for each request
-	client, scheme := newClientAndScheme(tlsConfig)
+func proxyAsync(engine *cluster.Engine, w http.ResponseWriter, r *http.Request, callback func(*http.Response)) error {
 	// RequestURI may not be sent to client
 	r.RequestURI = ""
 
+	client, scheme := engine.HTTPClientAndScheme()
```
Comment (dongluochen) [inline api/utils.go]: `client` could be nil in your definition. Should validate it before usage.

File: api/handlers.go [hunk 3]
```diff
@@ -462,7 +462,7 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	client, scheme := newClientAndScheme(c.tlsConfig)
+	client, scheme := container.Engine.HTTPClientAndScheme()
```
Comment (dongluochen) [inline api/handlers.go]: Need to validate `client`.

**Comment by nishanttotla** [comment]
@vieux is this related to #1549?

**Comment by vieux** [comment]
@nishanttotla yes !

**Comment by MHBauer** [comment]
Just to be clear, this makes it use the `samalba/dockerclient` instead of a raw golang http client. Right?

**Comment by vieux** [comment]
@MHBauer `samalba/dockerclient` uses the a  raw golang http client

Before this pr we were creating a new  raw golang http client for every proxy request,
now we simply reuse the same one as `dockerlient`, so we can reuse idle connection.

File: api/handlers.go (hunk 3 above)
Comment (vieux) [inline api/handlers.go]: added

File: api/utils.go (hunk 2 above)
Comment (vieux) [inline api/utils.go]: added

**Comment by dongluochen** [comment]
LGTM

File: api/handlers.go [hunk 4]
```diff
@@ -462,7 +462,11 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
...
+	if client == nil {
+		httpError(w, "cannot connect to docker engine", http.StatusInternalServerError)
```
Comment (nishanttotla) [inline api/handlers.go]: Small detail - is this the same error message as returned by engine on failed connection?

File: api/handlers.go [hunk 5]
```diff
@@ -767,7 +771,11 @@ func postContainersExec(c *context, w http.ResponseWriter, r *http.Request) {
...
```
Comment (nishanttotla) [inline api/handlers.go]: same here

File: api/handlers.go (hunk 4 above)
Comment (vieux) [inline api/handlers.go]: probably not, do you think it should ?

Comment (nishanttotla) [inline api/handlers.go]: I suppose it's not an issue. Perhaps just using caps cannot -> Cannot would be nice.

**Comment by abronan** [comment]
LGTM

**Pull request merged**